### PR TITLE
new post: This Week in Glean: rustc, iOS and an M1

### PR DESCRIPTION
**Pull request by badboy**
[Rendered](https://github.com/badboy/fnordig.de/blob/m1-twig/_posts/2021-04-16-rustc-ios-and-an-m1.md)

### CODE REVIEW

**Review by Dexterp37 — APPROVED** [review]
The Carthage fun seems fun..

File: _posts/2021-04-16-rustc-ios-and-an-m1.md [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+This article is [cross-posted on the Mozilla Data blog]().
+
+---
+
+Back in February I got an M1 MacBook.
+That's Apple's new ARM-based hardware.
+
+I got it with the explicit task to ensure that we are able to develop and build [Glean] on it.
+We maintain a Swift language binding, targeting iOS, and that one is used in [Firefox iOS]
+and also continue supporting development of the [Swift language bindings][swift lb], targeting iOS.
```
Comment (brizental) [inline _posts/2021-04-16-rustc-ios-and-an-m1.md]: This whole sentence is a bit confusing, I don't understand what you are trying to say.

File: _posts/2021-04-16-rustc-ios-and-an-m1.md [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+You still can't put code compiled for `arm64-apple-ios` and `arm64-apple-ios14.0-simulator` into the same universal binary[^5],
+because you can have each architecture only once (the `arm64` part in there).
+That's what Carthage and others stumbled over.
+
+Again Apple prepared for that and for a long time they want you to use [XCFramework bundles][XCFramework bundles][^6].
+Carthage just didn't used to support that.
+The [0.37.0 release][carthage release] fixed that.
+
+That still leaves Rust behind, as it doesn't know the new `-simulator` target.
+But as always the Rust community is ahead of the game and [deg4uss3r] started adding a new target in in [Rust PR #81966][rust sim target].
```
Comment (brizental) [inline _posts/2021-04-16-rustc-ios-and-an-m1.md]:
```suggestion
But as always the Rust community is ahead of the game and [deg4uss3r] started adding a new target in [Rust PR #81966][rust sim target].
```

File: _posts/2021-04-16-rustc-ios-and-an-m1.md [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+This was marked a stretch goal for a reason.
+This involves all the other teams with Rust code and the iOS teams too.
+We're not there yet and there's currently no explicit priority to make development of Firefox iOS on M1 hardware possible.
+But when it comes to it, Glean will be ready for it and the team can assist others to get it over the finish line.
+
+---
+
+Want to hear more about Glean and our cross-platform Rust development?
+Come to next week's [Rust Linz meetup][Rust Linz], where I will be talking about this.
```
Comment (brizental) [inline _posts/2021-04-16-rustc-ios-and-an-m1.md:144]: Nice! :D

**Review by brizental — APPROVED** [review]

File: _posts/2021-04-16-rustc-ios-and-an-m1.md (hunk 1 above)
Comment (badboy) [reply _posts/2021-04-16-rustc-ios-and-an-m1.md]: Oh, yeah, re-reading that it looks like some editing gone wrong, mixing two previous sentences. Fixing!

File: _posts/2021-04-16-rustc-ios-and-an-m1.md [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+Glean Python just ... worked.
+We use `cffi` to load the native library into Python.
+It gained `aarch64`[^2] macOS support in [v14.4.1][cffi].
+My colleague glandium later [contributed support code][arm64-wheels] so we build release wheels for that target too.
+So it's both possible to develop & test Glean Python, as well as use it as a dependency without having a full Rust development environment around.
+
+Glean Android is not that straight forward.
+Some of our transitive dependencies are based on years-old pre-built binaries of SQLite
+and of course there's not much support behind updating those Java libraries.
+It's possible, a friend managed to compile and run that library on an M1.
```
Comment (chutten) [inline _posts/2021-04-16-rustc-ios-and-an-m1.md]:
```suggestion
It's possible. A friend managed to compile and run that library on an M1.
```
(These are two independent clauses, so this is what's known as a "comma splice" which isn't recommended form.)

File: _posts/2021-04-16-rustc-ios-and-an-m1.md [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+But for Glean development we simply recommend relying on Rosetta 2, the x86_64 compatibility layer for now.
```
Comment (chutten) [inline _posts/2021-04-16-rustc-ios-and-an-m1.md]:
```suggestion
But for Glean development we simply recommend relying on Rosetta 2 (the x86_64 compatibility layer) for now.
```

File: _posts/2021-04-16-rustc-ios-and-an-m1.md [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+Now Xcode and the compiler will get confused what to put where when building on M1 hardware for both iOS hardware and the host architecture.
+
+So the compiler toolchain gained knowledge of a new thing: `arm64-apple-ios14.0-simulator`,
+explicitly marking the simulator target.
+The compiler knows from where to pick the libraries and other SDK files when using that target.
+You still can't put code compiled for `arm64-apple-ios` and `arm64-apple-ios14.0-simulator` into the same universal binary[^5],
+because you can have each architecture only once (the `arm64` part in there).
+That's what Carthage and others stumbled over.
+
+Again Apple prepared for that and for a long time they want you to use [XCFramework bundles][XCFramework bundles][^6].
```
Comment (chutten) [inline _posts/2021-04-16-rustc-ios-and-an-m1.md]:
```suggestion
Again Apple prepared for that and for a long time they have wanted you to use [XCFramework bundles][XCFramework bundles][^6].
```

File: _posts/2021-04-16-rustc-ios-and-an-m1.md [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+Carthage just didn't used to support that.
+The [0.37.0 release][carthage release] fixed that.
+
+That still leaves Rust behind, as it doesn't know the new `-simulator` target.
+But as always the Rust community is ahead of the game and [deg4uss3r] started adding a new target in [Rust PR #81966][rust sim target].
+He got half way there when I jumped in to push it over the finish line.
+How these targets work and how LLVM picks the right things to put into the compiled artifacts is severly underdocumented,
```
Comment (chutten) [inline _posts/2021-04-16-rustc-ios-and-an-m1.md:111]: Is there work to improve the documentation?

File: _posts/2021-04-16-rustc-ios-and-an-m1.md [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+Turns out: the `14.0` in `arm64-apple-ios14.0-simulator` is actually important.
+
+With the last missing piece in place, the new Rust target landed in February and is available in Nightly.
+Contrary to the main `aarch64-apple-darwin` or `aarch64-apple-ios` target, the simulator target is not Tier 2 yet
+and thus no prebuilt support is available.
+`rustup target add aarch64-apple-darwin` does _not_ work right now.
+I am now in discussions to [promote it to Tier 2][rust tier2 promotion],
+but it's currently blocked by the [RFC: Target Tier Policy].
+
+It works on nightly however and in combination with another cargo capability I'm able to build libraries for the M1 iOS simulator:
```
Comment (chutten) [inline _posts/2021-04-16-rustc-ios-and-an-m1.md]: "another cargo capability" makes me wonder _what_ cargo capability

File: _posts/2021-04-16-rustc-ios-and-an-m1.md [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+Come to next week's [Rust Linz meetup][Rust Linz], where I will be talking about this.
+
+---
+
+_Footnotes:_
+
+[^1]: See [Platform Support] for what the Tiers means.  
+[^2]: The other name for that target.  
+[^3]: "Apple Silicon" is yet another name for what is essentially the same as "M1" or "macOS aarch64"  
+[^4]: Or `arm64` for that matter. Yes, yet another name for the same thing.
```
Comment (chutten) [inline _posts/2021-04-16-rustc-ios-and-an-m1.md:153]: Might be fun to list all the "functionally identical, but syntactically distinct" ways of saying the same things that you discovered in this journey. 

It'd certainly make a fun slide in a talk.

File: _posts/2021-04-16-rustc-ios-and-an-m1.md [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+
+---
+
+_Footnotes:_
+
+[^1]: See [Platform Support] for what the Tiers means.  
+[^2]: The other name for that target.  
+[^3]: "Apple Silicon" is yet another name for what is essentially the same as "M1" or "macOS aarch64"  
+[^4]: Or `arm64` for that matter. Yes, yet another name for the same thing.  
+[^5]: ["Universal Binaries"][apple universal binary] exist for a long time now and allow for one binary to include the compiled artifacts for multiple targets. It's how there's only one Firefox for Mac download which runs natively on either Mac platform.
```
Comment (chutten) [inline _posts/2021-04-16-rustc-ios-and-an-m1.md]:
```suggestion
[^5]: ["Universal Binaries"][apple universal binary] have existed for a long time now and allow for one binary to include the compiled artifacts for multiple targets. It's how there's only one Firefox for Mac download which runs natively on either Mac platform.  
```

**Review by chutten — APPROVED** [review]
The footnotes don't render in the preview. They render fine when published, right?

File: _posts/2021-04-16-rustc-ios-and-an-m1.md (hunk 7 above)
Comment (badboy) [reply _posts/2021-04-16-rustc-ios-and-an-m1.md:111]: I didn't look much further, LLVM is _huge_ and I have no time to get familiar with their stuff.

File: _posts/2021-04-16-rustc-ios-and-an-m1.md (hunk 9 above)
Comment (badboy) [reply _posts/2021-04-16-rustc-ios-and-an-m1.md:153]: https://asahilinux.org/2021/03/progress-report-january-february-2021/#quick-terminology-note

**Review by travis79 — COMMENTED** [review]
🎉 Awesome, as usual :)

**Pull request merged**